src/util/kaldi-table-inl.h

```diff
@@ -178,25 +191,40 @@ template<class Holder>  class SequentialTableReaderScriptImpl:
       else  // orig_state_ was kLoadFailed, which only could have happened
         // if the user called FreeCurrent().
         KALDI_ERR << "You called Value() after FreeCurrent().";
-    } else if (state_ != kLoadSucceeded) {
+    } else if (state_ == kRangeExtractionFailed) {
+      KALDI_ERR << "Failed to load object from "
+                << PrintableRxfilename(data_rxfilename_) 
+                << " [" << range_ << "] "
+                << " (to suppress this error, add the permissive "
+                << "(p, ) option to the rspecifier.";
+    } else if (state_ != kLoadSucceeded && state_ != kRangeExtracted) {
       // This would be a coding error.
       KALDI_ERR << "Value() called at the wrong time.";
     }
-    return holder_.Value();
+    if (state_ == kRangeExtracted) {
+      return object_;
+    } else {
+      // Here state_ is kLoadSucceeded.
+      return holder_.Value();
+    }
   }
+
   void FreeCurrent() {
-    if (state_ == kLoadSucceeded) {
+    if (state_ == kLoadSucceeded || 
+        state_ == kRangeExtracted || state_ == kRangeExtractionFailed) {
       holder_.Clear();
       state_ = kLoadFailed;
     } else {
       KALDI_WARN << "FreeCurrent called at the wrong time.";
     }
   }
+
   void SwapHolder(Holder *other_holder) {
     // call Value() to ensure we have a value, and ignore its return value while
     // suppressing compiler warnings by casting to void.
     (void) Value();
-    if (state_ == kLoadSucceeded) {
+    if (state_ == kLoadSucceeded ||
+        state_ == kRangeExtracted || state_ == kRangeExtractionFailed) {
       holder_.Swap(other_holder);
```

Comment:
what's "bg" feature? Otherwise it's pretty clear to me. will do as you
suggested.

On Sun, Mar 27, 2016 at 12:26 AM, Daniel Povey <notifications@github.com>
wrote:

> In src/util/kaldi-table-inl.h
> <https://github.com/kaldi-asr/kaldi/pull/611#discussion_r57523237>:
>
> >    void SwapHolder(Holder *other_holder) {
> >      // call Value() to ensure we have a value, and ignore its return value while
> >      // suppressing compiler warnings by casting to void.
> >      (void) Value();
> > -    if (state_ == kLoadSucceeded) {
> > +    if (state_ == kLoadSucceeded ||
> > +        state_ == kRangeExtracted || state_ == kRangeExtractionFailed) {
> >        holder_.Swap(other_holder);
>
> Now that I look at this I realize there is an incompatibility (my fault)
> between the design of this feature, and the design of the "bg" feature.
> 'other_holder' needs to get the same value that we would have returned from
> Value(), otherwise the combination of these two features won't work. But
> this is not easy because there is no swap method in Holder that swaps with
> a raw object [and implementing such a thing in general would be hard, as we
> can't rely on Swap() functions always existing even for Kaldi objects].
>
> After thinking about this a bit, I think the best way to resolve this is a
> slight redesign of this object-indexing feature.
> Instead of a namespace function ExtractObjectRange, there will be such a
> function ExtractRange in all the Holder classes, and the type written to
> will be another Holder, and the 'object_' will instead be a holder object
> of type range_holder_. In most cases you'll give it an implementation that
> just throws.
>
> I think actually the easiest way to implement this in such a way that it
> has a functioning implementation for matrices, is to *keep* the existing
> ExtractObjectRange function that you've written, and have it called from
> the ExtractRange function of the KaldiObjectHolder. (BTW, for other Holder
> types you will have to write an ExtractRange function that throws; make
> sure to compile at src/ level to be sure you haven't missed any.)
>
> This redesign actually has another advantage. Currently we have no way to
> clear the object (since operator = is not guaranteed to exist), so when we
> do NextScpLine() we have no way to free memory tied up in the range object.
> Now that it's a Holder object, the memory can be freed by creating a
> temporary new Holder object, and swapping with the range_holder_.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/611/files/d778589ed02590c92d5578281dd181f22280fa8b#r57523237>
>



-- 
Xiao-hui